cowbird/permissions_synchronizer.py

```diff
@@ -74,59 +74,178 @@ def __init__(self,
+    def find_permissions_to_sync(self, permission, res_root_key):
+        # type: (Permission, str) -> Generator[Tuple[str, str], None, None]
         """
         Search and yield for every match a (service, permission name) tuple that is mapped with this permission.
         """
-        # check if the permission name is covered by this sync point
-        if not self.resource_match(permission):
-            return
-
         # For each permission mapping
         for mapping in self.mapping:
             # Does the current service has some mapped permissions?
-            if permission.service_name not in mapping:
+            if res_root_key not in mapping:
                 continue
             # Does the current permission access is covered?
-            if permission.name not in mapping[permission.service_name]:
+            if permission.name not in mapping[res_root_key]:
                 continue
             # This permission is mapped : yields matches
-            for svc, mapped_perm_name in mapping.items():
+            for res, mapped_perm_name in mapping.items():
                 # Don't synch with itself
-                if svc == permission.service_name:
+                if res == res_root_key:
                     continue
                 for perm_name in mapped_perm_name:
-                    yield svc, perm_name
+                    yield res, perm_name
 
-    def sync(self, perm_operation, permission):
-        # type: (Callable[Permission], Permission) -> None
+    def _find_matching_res(self, service_name, resource_nametype_path):
+        # type: (str, str) -> (str, int)
+        """
+        Finds a resource key that matches the input resource path, in the sync_permissions config.
+        Note that it returns the longest match and only the named segments of the path are included in the length value.
+        Any tokenized segment is ignored in the length.
+
+        @param service_name Name of the service associated with the input resource.
+        @param resource_nametype_path Full resource path name, which includes the type of each segment
+                                       (ex.: /name1:type1/name2:type2)
+        """
+        service_resources = self.services[service_name]
+        # Find which resource from the config matches with the input permission's resource tree
+        # The length of a match is determined by the number of named segments matching the input resource.
+        # MULTI_TOKEN name tokens can match the related type 0 to N times. They are ignored from the match length since
+        # it is ambiguous in the case of a 0 length match.
+        # SINGLE_TOKEN name tokens can match exactly 1 time only. They are ignored from the match length since we favor
+        # a match with specific names over another with SINGLE_TOKEN generic tokens.
+        matched_res_dict = {}
+        for res_key, res_segments in service_resources.items():
+            match_len = 0
+            res_regex = "^"
```

Comment:
Is the `r` really useful here (and in other mentions)? Since the string does not contain any backslash, I considered it was not necessary here.

Reply:
I think its more of a convention I like to apply anytime regex are used to avoid switching between escaped or literal representations, but it technically doesn't change the interpretation here.